<font face='Liberation serif' size=3>
    
#### **Привет!** 👋

Меня зовут Артем Пиляй. Я буду проверять твой проект. В рамках корпоративной этики мы общаемся на «ты». Но если лучше на «Вы», то нужно обозначить это сразу.
 
В тетрадке есть комментарии к проекту, они будут в отдельных ячейках Markdown.
 
<div class="alert alert-danger">

**Артём:** ❌ Такой комментарий **нужно исправить** обязательно, он критически влияет на удачное выполнение проекта.

</div>

<div class="alert alert-success">

**Артём:** 👍 Это **главный комментарий**, значит все сделано хорошо. 😉 А еще в таких комментариях бывают подсказки как сделать **твой проект ещё лучше**.

</div>

---

**Распространенные ошибки**

<div class="alert alert-danger">

**Артём:** ❌ Не перемещай и не удаляй мои комментарии. Лучше, чтобы твои ответы на мои комментарии шли по очереди, как сообщения в мессенджере. Это ускорит проверку.
    
</div>

<div class="alert alert-danger">

**Артём:** ❌ Проверь проект на работоспособность перед отправкой на проверку. Это важный навык, который тебе пригодится как на собеседовании, при выполнении тестового задания, так и перед совещанием у руководства компании. Помни, ревьеры не вносят изменения в код, даже если это путь к датасету или ячейка в неверном формате (code/mardown).  
    
</div>

<div class="alert alert-success">

**Артём:** 👍 Кстати, смотри как можно сделать, чтобы датасет грузился и локально, и в JupyterHub:

```py
try:
    df = pd.read_csv('C:/Users/Student/Desktop/dataset.csv')
except:
    df = pd.read_csv('/datasets/dataset.csv')
```
    
Используя эту конструкцию в будущих проектах, ты будешь спокойно отправлять проект на проверку зная, что он будет работать.
    
</div>

<div class="alert alert-danger">

**Артём:** ❌ Проект должен быть структурирован. Где названия разделов и подразделов будут записаны в ячейках [Markdown](https://paulradzkov.com/2014/markdown_cheatsheet/?).
    
</div>

<div class="alert alert-danger">

**Артём:** ❌ Если с твоим проектом будет знакомиться другой человек, например рекрутер или старший дата специалист, в нем должны быть описаны поставленные задачи, содержание датасета, промежуточные и финальные выводы. Как правило, заказчик читает только финальный вывод, поэтому важно добавить информацию о проведенном исследовании, применить форматирование текста, в общем, сделать красиво, чтобы было понятно, что проведено настоящее исследование и выводу можно верить. 

</div>  

<div class="alert alert-success">

**Артём:** 👍 Длинный код и комментарии нужно переносить на новую строку. Согласись, коллеги не оценят, если им придется тратить время на пролистывание ячеек. Согласно международному стандарту [PEP8](https://www.python.org/dev/peps/pep-0008/) длина строки не должна превышать 79 символов.
Вот как можно перенести код на новую строку:

```py
# Переносить можно код внутри скобки
example = (2 *
           2)
```
Конечно, я не считаю сколько символов в строке, главное, чтобы код или комментарий не ушел за пределы ячейки (зависит от разрешения экрана). Если ушел, то это желтый комментарий для каждой нечитаемой ячейки.
    
</div>

<div class="alert alert-success">

**Артём:** 👍 Тимлид, старший дата специалист или ревьюер ценят твое и своё время, подсвечивай свои комментарии в ячейках Markdown. Шаблон я тебе оставил ниже. Подробнее про синтаксис Markdown, с примерами, можно почитать [здесь](https://paulradzkov.com/2014/markdown_cheatsheet/?).
    
</div>

<div class="alert alert-success">

**Артём:** 👍 Буду рад, если воспользуешься моими советами. Они тебе помогут при трудоустройстве, например в [Яндексе](https://yandex.ru/jobs/pages/analytics_interview?ysclid=l84s63fs1g790141490), и при работе в команде. Если в ходе проведения исследования возникнут вопросы, не стесняйся задавай. 🙂
</div>

<div class="alert alert-info">

**Артём:** 🏁 [Итоги ревью здесь](#results)
</div>


**Поехали!** 🚀 👨‍🚀
<a id="start"></a>

<div style="background-color:lavender; color:#black; text-align:left; padding:2%">
    
#### **Студент(ка):** ✅ Это шаблон для тебя. Твой комментарий или вопрос в такой ячейке я точнно увижу.

# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

,children,days_employed,dob_years,education,education_id,family_status,family_status_id,gender,income_type,debt,total_income,purpose
0,1,-8437.673028,42,высшее,0,женат / замужем,0,F,сотрудник,0,253875.639453,покупка жилья
1,1,-4024.803754,36,среднее,1,женат / замужем,0,F,сотрудник,0,112080.014102,приобретение автомобиля
2,0,-5623.422610,33,Среднее,1,женат / замужем,0,M,сотрудник,0,145885.952297,покупка жилья
3,3,-4124.747207,32,среднее,1,женат / замужем,0,M,сотрудник,0,267628.550329,дополнительное образование
4,0,340266.072047,53,среднее,1,гражданский брак,1,F,пенсионер,0,158616.077870,сыграть свадьбу
5,0,-926.185831,27,высшее,0,гражданский брак,1,M,компаньон,0,255763.565419,покупка жилья
6,0,-2879.202052,43,высшее,0,женат / замужем,0,F,компаньон,0,240525.971920,операции с жильем
7,0,-152.779569,50,СРЕДНЕЕ,1,женат / замужем,0,M,сотрудник,0,135823.934197,образование
8,2,-6929.865299,35,ВЫСШЕЕ,0,гражданский брак,1,F,сотрудник,0,95856.832424,на проведение свадьбы
9,0,-2188.756445,41,среднее,1,женат / замужем,0,M,сотрудник,0,144425.938277,покупка жилья для семьи


**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
#--
#Общее количество заемщиков в разрезе кол-ва детей
total_by_children = data.groupby('children')['debt'].count()
total_data = pd.DataFrame(
    {'children':total_by_children.index,
     'total':total_by_children.values})
print(total_data)

   children  total
0         0  14091
1         1   4808
2         2   2052
3         3    330
4         4     41
5         5      9


In [23]:
#Общее количество заемщиков c просроченной задолженностью в разрезе кол-ва детей
with_debt_by_children = data[data['debt'] == 1]
grouped_with_debt_by_children = with_debt_by_children.groupby('children')['debt'].count()
total_with_debt = pd.DataFrame(
    {'children':grouped_with_debt_by_children.index,
     'with_debt':grouped_with_debt_by_children.values})
    
print(total_with_debt)

   children  with_debt
0         0       1063
1         1        444
2         2        194
3         3         27
4         4          4


In [24]:
#объединение значений
res = total_data.merge(total_with_debt, on='children', how='left')
print(res)

   children  total  with_debt
0         0  14091     1063.0
1         1   4808      444.0
2         2   2052      194.0
3         3    330       27.0
4         4     41        4.0
5         5      9        NaN


In [25]:
#обработка пустых значений
res = res.fillna(0)
#преобразование значений with_debt в целые числа
res['with_debt'] = res['with_debt'].astype('int')
print(res)

   children  total  with_debt
0         0  14091       1063
1         1   4808        444
2         2   2052        194
3         3    330         27
4         4     41          4
5         5      9          0


In [26]:
#вычисление соотношения задолжников к общему количеству заемщиков
res['ratio'] = round(res['with_debt'] / res['total']*100, 2)
print(res)

   children  total  with_debt  ratio
0         0  14091       1063   7.54
1         1   4808        444   9.23
2         2   2052        194   9.45
3         3    330         27   8.18
4         4     41          4   9.76
5         5      9          0   0.00


<div class="alert alert-success">

**Артём:** 👍 Молодец, что используешь сводную таблицу.

</div>

<div class="alert alert-success">

#### **Артём:** 👍 При выводе на экран датафреймов лучше использовать метод `display()`, а не `print()`. Попробуй здесь и далее.
    
</div>

In [27]:
#вычисление соотношения задолжников к общему количеству заемщиков
#вывод результатов с помощью display
res['ratio'] = round(res['with_debt'] / res['total']*100, 2)
display(res)

,children,total,with_debt,ratio
0,0,14091,1063,7.54
1,1,4808,444,9.23
2,2,2052,194,9.45
3,3,330,27,8.18
4,4,41,4,9.76
5,5,9,0,0.00


<div style="background-color:lavender; color:#black; text-align:left; padding:2%">
    
Спасибо, так симпатичнее)

<div class="alert alert-success">

**Артём:** 👍 Ага, согласен.

</div>

**Вывод:** 

Из полученных результатов можно сделать следующие выводы:

- у заемщиков без детей вероятность возврата кредита в срок незначительно выше, чем у заемщиков с количеством детей от 1 до 4

- заемщики, у которых 5 детей, возвращают кредит в срок, но количество общего числа заемщиков с таким количеством детей намного ниже, чем заемщиков с меньшим количеством детей либо без детей.

<div class="alert alert-success">

**Артём:** 👍 Правильно, что смотришь на репрезентативность выборок.

</div>

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [28]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
#Общее количество заемщиков в разрезе семейного положения
total_by_family_status = data.groupby('family_status')['debt'].count()
total_data_by_family_status = pd.DataFrame(
    {'family_status':total_by_family_status.index,
     'total':total_by_family_status.values})
print(total_data_by_family_status)

           family_status  total
0  Не женат / не замужем   2796
1              в разводе   1189
2         вдовец / вдова    951
3       гражданский брак   4134
4        женат / замужем  12261


In [29]:
#Найдем общее количество должников в разрезе семейного положения
with_debt_by_family_status = data[data['debt'] == 1]
grouped_with_debt_by_family_status = with_debt_by_family_status.groupby('family_status')['debt'].count()
total_with_debt_by_family_status = pd.DataFrame(
    {'family_status':grouped_with_debt_by_family_status.index,
     'with_debt':grouped_with_debt_by_family_status.values})
    
print(total_with_debt_by_family_status)

           family_status  with_debt
0  Не женат / не замужем        273
1              в разводе         84
2         вдовец / вдова         63
3       гражданский брак        385
4        женат / замужем        927


In [30]:
#объединение значений
res_by_family_status = total_data_by_family_status.merge(
    total_with_debt_by_family_status,
    on='family_status', how='left')
print(res_by_family_status)

           family_status  total  with_debt
0  Не женат / не замужем   2796        273
1              в разводе   1189         84
2         вдовец / вдова    951         63
3       гражданский брак   4134        385
4        женат / замужем  12261        927


In [31]:
#Найдем соотношение задолжников к общему количеству заемщиков
res_by_family_status['ratio'] = round(
    res_by_family_status['with_debt'] / res_by_family_status['total']*100, 2)
display(res_by_family_status)

,family_status,total,with_debt,ratio
0,Не женат / не замужем,2796,273,9.76
1,в разводе,1189,84,7.06
2,вдовец / вдова,951,63,6.62
3,гражданский брак,4134,385,9.31
4,женат / замужем,12261,927,7.56


**Вывод:** 

Из полученных значений можно сделать следующие выводы:

- вне зависимости от семейного положения вероятность задержки возврата кредита в срок не превышает 10%
- наименьшая вероятность задержки возврата кредита в срок наблюдается у заемщиков, которые находятся в семейном положении "вдовец / вдова"
- наибольшая вероятность задержки кредита в срок наблюдается у заемщиков, которые находятся в семейном положении "Не женат / не замужем"

<div class="alert alert-success">

**Артём:** 👍 Верно.

</div>

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [32]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
#Общее количество заемщиков в разрезе категории уровня дохода
total_by_income_category = data.groupby('total_income_category')['debt'].count()
total_data_by_income_category = pd.DataFrame(
    {
        'income_category':total_by_income_category.index,
        'total':total_by_income_category.values
    })
print(total_data_by_income_category)

  income_category  total
0               A     25
1               B   5014
2               C  15921
3               D    349
4               E     22


In [33]:
#Найдем общее количество должников в разрезе категории уровня дохода
with_debt_by_income_category = data[data['debt'] == 1]
grouped_with_debt_by_income_category = with_debt_by_income_category.groupby('total_income_category')['debt'].count()
total_with_debt_by_income_category = pd.DataFrame(
    {
        'income_category':grouped_with_debt_by_income_category.index,
        'with_debt':grouped_with_debt_by_income_category.values
    })
    
print(total_with_debt_by_income_category)

  income_category  with_debt
0               A          2
1               B        354
2               C       1353
3               D         21
4               E          2


In [34]:
#Объединим полученные значения в одну таблицу
res_by_income_category = total_data_by_income_category.merge(
    total_with_debt_by_income_category, 
    on='income_category', how='left')
print(res_by_income_category)

  income_category  total  with_debt
0               A     25          2
1               B   5014        354
2               C  15921       1353
3               D    349         21
4               E     22          2


In [35]:
#Найдем соотношение задолжников к общему количеству заемщиков
res_by_income_category['ratio'] = round(
    res_by_income_category['with_debt'] / res_by_income_category['total']*100, 2)
display(res_by_income_category)

,income_category,total,with_debt,ratio
0,A,25,2,8.00
1,B,5014,354,7.06
2,C,15921,1353,8.50
3,D,349,21,6.02
4,E,22,2,9.09


**Вывод:** 

Из полученных значений можно сделать следующие выводы:

- вне зависимости от категории уровня дохода вероятность задержки возврата кредита в срок не превышает 10%
- наименьшая вероятность задержки возврата кредита в срок наблюдается у заемщиков с категорием дохода = D, то есть с уровнем дохода 30001–50000
- наибольшая вероятность задержки возврата кредита в срок наблюдается у заемщиков с категорией дохода = E , то есть с уровнем дохода 0–30000

<div class="alert alert-success">

#### **Артём:** 👍 Дам тебе совет на будущее, как лучше проводить категоризацию количественных данных. **В этом проекте переделывать ничего не надо, так как заказчик уже предложил свои категории, его стоит проинформировать, что выборки крайне несбалансированны.** Если только как дополнительное исследование.
    
Сейчас самая маленькая категория в 728 раз меньше самой большой. Это может отразиться на результатах исследования. Когда мы делим на категории количественные данные, то лучше их делать примерно одинаковыми. Для этого можно воспользоваться методом `qcut()`. Пример:
    
```py
pd.qcut(df['total_income'], 4)
```
    
```
Name: total_income, Length: 21453, dtype: category
Categories (4, interval[float64]): [(20666.999, 107620.0]  (107620.0, 142594.0]  (142594.0, 195818.0]  (195818.0, 2265604.0]]
```
                                                                                                               
Что внутри: 
    
```py
pd.qcut(датафрейм['столбец'], количество_категорий)
``` 
    
Итого у нас 4 категории:
<ul>
  <li>меньше 107_620</li>
  <li>от 107_620 до 142_594</li>
  <li>от 142_594 до 195_818</li>
  <li>от 195_818</li>
    
Пример использования:
```py    
# Создаем категории дохода с помощью pd.qcut
data['total_qcut_category'] = pd.qcut(data['total_income'], 5, labels=['A', 'B', 'C', 'D', 'E'])

# Группируем данные по новому столбцу 'total_income_category'
pivot_table = data.groupby('total_qcut_category').agg(
    borrowers=('debt', 'size'),
    debtors=('debt', 'sum')
).reset_index()
```    
    
</ul>

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [36]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
#Общее количество заемщиков в разрезе категории цели кредита
total_by_purpose_category = data.groupby('purpose_category')['debt'].count()
total_data_by_purpose_category = pd.DataFrame(
    {'purpose_category':total_by_purpose_category.index,
     'total':total_by_purpose_category.values})
print(total_data_by_purpose_category)

           purpose_category  total
0    операции с автомобилем   4279
1  операции с недвижимостью  10751
2     получение образования   3988
3        проведение свадьбы   2313


In [37]:
with_debt_by_purpose_category = data[data['debt'] == 1]
grouped_with_debt_by_purpose_category = with_debt_by_purpose_category.groupby('purpose_category')['debt'].count()
total_with_debt_by_purpose_category = pd.DataFrame(
    {'purpose_category':grouped_with_debt_by_purpose_category.index,
     'with_debt':grouped_with_debt_by_purpose_category.values})
    
print(total_with_debt_by_purpose_category)

           purpose_category  with_debt
0    операции с автомобилем        400
1  операции с недвижимостью        780
2     получение образования        369
3        проведение свадьбы        183


In [38]:
#Объединим полученные значения в одну таблицу
res_by_purpose_category = total_data_by_purpose_category.merge(
    total_with_debt_by_purpose_category,
    on='purpose_category', how='left')
print(res_by_purpose_category)

           purpose_category  total  with_debt
0    операции с автомобилем   4279        400
1  операции с недвижимостью  10751        780
2     получение образования   3988        369
3        проведение свадьбы   2313        183


In [39]:
#Найдем соотношение задолжников к общему количеству заемщиков
res_by_purpose_category['ratio'] = round(
    res_by_purpose_category['with_debt'] / res_by_purpose_category['total']*100, 2)
print(res_by_purpose_category)

           purpose_category  total  with_debt  ratio
0    операции с автомобилем   4279        400   9.35
1  операции с недвижимостью  10751        780   7.26
2     получение образования   3988        369   9.25
3        проведение свадьбы   2313        183   7.91


**Вывод:** 

Из полученных значений можно сделать следующие выводы:

- вне зависимости от цели кредита вероятность задержки возврата кредита в срок не превышает 10%
- наименьшая вероятность задержки возврата кредита в срок наблюдается у заемщиков с целью кредита = "операции с недвижимостью"
- наибольшая вероятность задержки возврата кредита в срок наблюдается у заемщиков с целями кредита = "операции с автомобилем" и "получение образования"

<div class="alert alert-success">

**Артём:** 👍 Правильно.

</div>

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* 
Возможные причины:

- данные атрибуты были необязательными при заполнении анкеты заемщика
- техническая ошибка при выгрузке данных

<div class="alert alert-success">

**Артём:** 👍 Все верно, технический и человеческий факторы.

</div>

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* поскольку среднее значение может некорректно характеризовать данные, если среди них есть слишком выдающиеся значения. В то время как медианные значения характеризуют значения, которое находятся по центру списка чисел в наборе упорядоченных данных.

<div class="alert alert-success">

**Артём:** 👍 Молодец. Действительно важно учитывать, что в данных могут быть выбросы, которые могут сместить итоговый результат если использовать при заполнении данных среднее значение.

</div>

### Шаг 4: общий вывод.

Вне зависимости от рассматриваемых категорий заемщиков, вероятность задержки возврата кредита в срок не превышает 10%.

<div class="alert alert-success">

**Артём:**  Промежуточные выводы будет смотреть старший аналитик, а заказчик будет читать финальный вывод и только его. Добавь информации о полученных результатах по каждому из четырех вопросов, разбей по пунктам, сделай красиво, чтобы было понятно, что проведено настоящее исследование и выводу можно верить. Например, какие именно категории имеют наибольший процент невозврата кредитов, и почему? Учитывай для кого проведено исследование, в банковской сфере даже сотые доли процента играют роль, но в нашем можно ограничиться десятыми. Помни, вывод – это лицо дата специалиста. 😉

</div>

<div style="background-color:lavender; color:#black; text-align:left; padding:2%">
    
**Развернутый вывод:**
    
В результате проведенных исследований получены следующие выводы:

1. **В зависимости от количества детей:**
    
    - заемщики, у которых 5 детей, возвращают кредит в срок. При этом данная група заемщиков является наименьшей по количеству среди рассматриваемых в разрезе количества детей, и составляет 9 заемщиков;
    
    - у заемщиков без детей вероятность задержки возврата кредита в срок составляет 7.54%. Данная категория заемщиков является наибольшей по количеству взятых кредитов и составляет 14091 заемщика;
    
    - у заемщиков, у которых количество детей составляет от 1 до 4, вероятность задержки возврата кредита в срок составляет от 8.18% до 9.76%. Наибольшее количество кредитов среди данной выборки у заемщиков с 1 ребенком, а вероятность задержки возврата кредита в срок составляет 9.23%.
    
| Количество детей| Всего заемщиков | Из них с задержкой возврата | Соотношение |
|:--------------- |:---------------:| :--------------------------:|------------:|
| 0               | 14091           | 1063                        |7.54         |
| 1               | 4808            | 444                         |9.23         |
| 2               | 2052            | 194                         |9.45         |
| 3               | 330             | 27                          |8.18         |
| 4               | 41              | 4                           |9.76         |
| 5               | 9               | 0                           |0.00         |
    
    
    
2. **В зависимости от семейного положения:**
    
    - вне зависимости от семейного положения вероятность задержки возврата кредита в срок не превышает 10%;
    
    - наименьшая вероятность задержки возврата кредита в срок наблюдается у заемщиков, которые находятся в семейном положении "вдовец / вдова", и составляет 6.62%. При этом данная група заемщиков является наименьшей по количеству среди рассматриваемых в разрезе семейного положения, и составляеа 951 заемщика;

    - наибольшая вероятность задержки кредита в срок наблюдается у заемщиков, которые находятся в семейном положении "Не женат / не замужем", и составляет 9.76%.
    
    - наибольшее количество кредитов среди данной выборки наблюдается у заемщиков, которые находятся в семейном положении "женат / замужем", данная группа составляет 12261 заемщика. Вероятность задержки возврата кредита в срок у данной группы составляет 7.56%.
    
| Семейное положение   | Всего заемщиков | Из них с задержкой возврата | Соотношение |
|:-------------------- |:---------------:| :--------------------------:|------------:|
| Не женат / не замужем| 2796            | 273                         |9.76         |
| В разводе            | 1189            | 84                          |7.06         |
| Вдовец / вдова       | 951             | 63                          |6.62         |
| Гражданский брак     | 4134            | 385                         |9.31         |
| Женат / замужем      | 12261           | 927                         |7.56         |
    
    
3. **В зависимости от уровня дохода:**
    
    - вне зависимости от категории уровня дохода вероятность задержки возврата кредита в срок не превышает 10% для каждой категории;

    - наименьшая вероятность задержки возврата кредита в срок наблюдается у заемщиков с категорией дохода = D, то есть с уровнем дохода 30001–50000.

    - наибольшая вероятность задержки возврата кредита в срок наблюдается у заемщиков с категорией дохода = E, то есть с уровнем дохода 0–30000. При этом, данная выборка является самой немногочисленной по количеству взятых кредитов, и составляет 22 заемщика;
    
    - наибольшее количество взятых кредитов наблюдается в категории дохода = C, то есть у заемщиков с уровнем дохода 50001–200000, и составляет 15921 заемщика. Вероятность задержки возврата кредита у таких заемщиков составляет 8.50%.
    
| Уровень дохода | Всего заемщиков | Из них с задержкой возврата | Соотношение |
|:-------------- |:---------------:| :--------------------------:|------------:|
| 1000001 и выше | 25              | 2                           |8.00         |
| 200001–1000000 | 5014            | 354                         |7.06         |
| 50001–200000   | 15921           | 1353                        |8.50         |
| 30001–50000    | 349             | 21                          |6.02         |
| 0–30000        | 22              | 2                           |9.09         |

    
    
4. **В зависимости от цели кредита:**
    
    - вне зависимости от цели кредита вероятность задержки возврата кредита в срок не превышает 10%;
    
    - наименьшая вероятность задержки возврата кредита в срок наблюдается у заемщиков с целью кредита = "операции с недвижимостью". Данная категория заемщиков так же являеься наиболее многочисленный по количеству взятых кредитов и составляет 10751 заемщик;
    
    - наибольшая вероятность задержки возврата кредита в срок наблюдается у заемщиков с целями кредита = "операции с автомобилем" и "получение образования", и составляет 9.35% и 9.25% соответственно. 
    
| Цель кредита             | Всего заемщиков | Из них с задержкой возврата | Соотношение |
|:------------------------ |:---------------:| :--------------------------:|------------:|
| операции с автомобилем   | 4279            | 400                         |9.35         |
| операции с недвижимостью | 10751           | 780                         |7.26         |
| получение образования    | 3988            | 369                         |9.25         |
| проведение свадьбы       | 2313            | 183                         |7.91         |
    

<div class="alert alert-success">

**Артём:** 👍 Отличный вывод. Аккуратно, последовательно приятно читать.

</div>

## 🏁 Итоги ревью
<a id="results"></a>
[В начало](#start)


<div class="alert alert-danger">

**Артём:** ❌ Алексей, спасибо за работу. Осталось дополнить финальный вывод и можно идти к заказчику. Молодец! 👍
    
Для удобства, комментарии которые нужно скорректировать я добавил в меню **Contents** слева.
    
Жду новую версию проекта. 🙂

</div>

<img src="https://s19536.pcdn.co/wp-content/uploads/2020/06/Web-Great-Job.jpg"> 
<div class="alert alert-success">   

#### **Артём:** ✌️ Алексей, спасибо за работу и внесенные корректировки! Критических замечаний нет.
    

Поздравляю, **проект принят!** 🥳 🎉 Ты молодец!

Желаю тебе успехов в дальнейшем обучении. 😉
    
---
    
#### 📖 **Полезные ссылки**

* [Работа в Яндексе](https://yandex.ru/jobs/pages/analytics_interview?ysclid=l84s63fs1g790141490)
* [Настройте свой Jupyter Notebook правильно](https://medium.com/nuances-of-programming/настройте-свой-jupyter-notebook-правильно-faef14c89434)    
* [Краткое руководство по Маркдауну](https://paulradzkov.com/2014/markdown_cheatsheet/?)
* [PEP 8 – Style Guide for Python Code](https://www.python.org/dev/peps/pep-0008/)    
* [Обработка пропусков в данных](https://loginom.ru/blog/missing)  
* [Разделение данных с помощью qcut и cut в Pandas](https://dfedorov.spb.ru/pandas/Разделение%20данных%20в%20Pandas%20с%20помощью%20qcut%20и%20cut.html?)  
* [Официальная документаци метода qcut](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.qcut.html)  
* [Подробное руководство по группировке и агрегированию с помощью pandas](https://dfedorov.spb.ru/pandas/Подробное%20руководство%20по%20группировке%20и%20агрегированию%20с%20помощью%20pandas.html?ysclid=lcmm6rtg2w602526506)
* [Проверка статистических гипотез. (21 мая 2020 г.)](https://allatambov.github.io/psms/pdf/hypo-test.pdf)    
* [Шпаргалка по визуализации данных в Python с помощью Plotly](https://habr.com/ru/articles/502958/) 
* [Тест гигиены визуализации данных](https://bar-or-pie.dianov.org)    
* [Задачки SQL](https://habr.com/ru/company/dcmiran/blog/500360/)    
* [7 cайтов для оттачивания навыка написания SQL запросов на 2020 год](https://techrocks.ru/2019/11/22/7-sites-to-perfect-sql-query-writing-skills/)
* [Главред помогает очистить текст от словесного мусора, проверяет на соответствие информационному стилю.](https://glvrd.ru)

</div>    